include/antler/project/object.hpp

```diff
@@ -95,7 +95,11 @@ namespace antler::project {
          const auto& itr = m_dependencies.find(dep.name());
          bool has_value = itr == m_dependencies.end();
 
-         m_dependencies.emplace(dep.name(), std::move(dep));
+         if(!has_value)
+            itr->second = dep;
+         else
+            m_dependencies.emplace(dep.name(), std::move(dep));
```

Comment:
Is this undefined behavior because we don't know what will be the order of evaluating arguments?
So first we might move `dep` and later read `dep.name()` which will be undefined?

Reply:
If this is undefined behavior, I think it's more like forwarding happens on name and move before construction of the key, val pair. So when the pair is created (the internal storage for unordered_map) name is no longer valid.

Did you try this with [UBsan](https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html#id3)?

Reply:
to my best knowledge this is UB as order of arguments evaluation of function is unspecified. So even if such thing work I wouldn't rely on it as on some compiler (or specific version of it) it could break.
If you want to deeper understand that, google for 'order of evaluation' and 'sequence points'. I know there are some changes in c++17 comparing to older standards.

Reply:
> Did you try [UBsan](https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html#id3)?
> 
> I think the process of emplace will do a forward of both args into the next function at essentially the same time. Not really an order of evaluation.

I couldn't compile whole code with clang++ (due to external projects), so I have extracted following code:
https://godbolt.org/z/4953rEfh5
and UBsan did not show anything.

Later I found this: https://stackoverflow.com/questions/27266755/is-it-safe-to-pass-as-argument-a-member-of-an-object-which-is-moving

and in short: in this case it is defined.
Anyway I will change it so no one else needs to think of it. New proposed version:

```suggestion
            auto name = dep.name();
            m_dependencies.emplace(std::move(name), std::move(dep));
```

Reply:
@mikelik nice investigation, thanks. I'm surprised `map` container internal structures such as `std::pair` are defined in standard. Anyway it doesn't seem to be reliable as we rely on implementation's internal order of initialization. Easier to make it straight than to ponder if it could break or not. This is not performance critical code anyway (although I doubt there is any performance downgrade as object will be moved anyway)